### PR TITLE
Plug memory leaks

### PR DESCRIPTION
**Pull request by gezalore**
We really should have an ASAN CI job, but unfortunately whenever we exit with an error it's a bit of a mess. Maybe we could run only those test in the CI that are not expected to fail?

### CODE REVIEW

**Comment by gezalore** [comment]
BTW there are plenty more, I just stopped there.

**Review by wsnyder — COMMENTED** [review]
Much appreciated.

I'd generally like to use only pushDeleteps if we can, otherwise there's a risk that fixing some of these leaks may expose another bug (that was hidden because of the lack of delete) if there's a dangling reference in a user() pointer to what's being deleted.

**Comment by gezalore** [comment]
Done. Note however that using pushDeletep blindly can cause problems with memory consumption. I pushed a fix to bit-op-tree recently that caused it to consume 500% extra memory (~50GB!) due to the deferred deletion. Obviously we could be more clever not creating redundant trees there, but that's another matter.

File: src/V3Dfg.h [hunk 1]
```diff
@@ -331,9 +331,11 @@ class DfgVertex VL_NOT_FINAL {
         // For simplicity, all packed types are represented with a fixed type
         if (AstUnpackArrayDType* const typep = VN_CAST(nodep->dtypep(), UnpackArrayDType)) {
             // TODO: these need interning via AstTypeTable otherwise they leak
-            return new AstUnpackArrayDType{typep->fileline(),
-                                           dtypeForWidth(typep->subDTypep()->width()),
-                                           typep->rangep()->cloneTree(false)};
+            AstNodeDType* const dtypep = new AstUnpackArrayDType{
```
Comment (wsnyder) [inline src/V3Dfg.h:334]: Remove TODO comment above.  I don't know why we didn't put it into the type table earlier.

File: src/V3OrderParallel.cpp [hunk 2]
```diff
@@ -1306,6 +1306,16 @@ class Contraction final {
...
             contract(mergeCanp);
         }
+
+        // Free remaining SiblingMCs
+        while (MergeCandidate* const mergeCanp = m_sb.best()) {
+            m_sb.remove(mergeCanp);
+            if (SiblingMC* const smcp = mergeCanp->toSiblingMC()) {
+                smcp->unlinkA();
+                smcp->unlinkB();
+                delete smcp;
```
Comment (wsnyder) [inline src/V3OrderParallel.cpp:1316]:
```suggestion
            if (SiblingMC* smcp = mergeCanp->toSiblingMC()) {
                smcp->unlinkA();
                smcp->unlinkB();
                VL_DO_DANGLING(delete smcp, smcp);
```

File: src/V3Task.cpp [hunk 3]
```diff
@@ -462,7 +462,7 @@ class TaskVisitor final : public VNVisitor {
             UINFO(9, "     Port " << portp << endl);
             UINFO(9, "      pin " << pinp << endl);
             if (inlineTask) {
-                pinp->unlinkFrBack();  // Relinked to assignment below
+                pushDeletep(pinp->unlinkFrBack());  // Cloned in assignment below
```
Comment (wsnyder) [inline src/V3Task.cpp:465]: Probably need VL_DO_DANGLING on the three deletes in this file (or comment otherwise).

File: src/V3ParseGrammar.cpp [hunk 4]
```diff
@@ -168,6 +168,11 @@ AstVar* V3ParseGrammar::createVariable(FileLine* fileline, const string& name,
     if (GRAMMARP->m_varIO == VDirection::NONE && GRAMMARP->m_varDecl == VVarType::PORT) {
         // Just a port list with variable name (not v2k format); AstPort already created
         if (dtypep) fileline->v3warn(E_UNSUPPORTED, "Unsupported: Ranges ignored in port-lists");
+        if (arrayp) VL_DO_DANGLING(arrayp->deleteTree(), arrayp);
+        if (attrsp) {
+            // TODO: Warn that attribute is ignored here?
```
Comment (wsnyder) [inline src/V3ParseGrammar.cpp:173]:
```suggestion
            // TODO: Merge attributes across the list? Or warn attrs ignored.
```

File: src/V3Broken.cpp [hunk 5]
```diff
@@ -372,7 +372,9 @@ void V3Broken::selfTest() {
     // Exercise addNewed and deleted for coverage, as otherwise only used with VL_LEAK_CHECKS
     FileLine* const fl = new FileLine{FileLine::commandLineFilename()};
     const AstNode* const newp = new AstBegin{fl, "[EditWrapper]", nullptr};
+#ifndef VL_LEAK_CHECKS
```
Comment (wsnyder) [inline src/V3Broken.cpp:375]: Comment as to why.

File: src/V3Task.cpp (hunk 3 above)
Comment (gezalore) [reply src/V3Task.cpp:465]: They are all used later, but the control flow is a mess, so I disn't want to do

if (a) foo->replaceWith(bar)
dosomethingwith(foo)
if (a) pushDeleteo(foo)

Added comments

File: src/V3ParseGrammar.cpp (hunk 4 above)
Comment (gezalore) [reply src/V3ParseGrammar.cpp:173]: FWIW this is for 

```
module (clk /* verilator clocker */ <- igonred
);

input clk
endmodule
```
in t_clk_concat_2

**Comment by gezalore** [comment]
Done

**Review by wsnyder — APPROVED** [review]

**Pull request merged**